CMakeLists.txt

```diff
@@ -48,19 +48,134 @@ if(POLICY CMP0135)
   cmake_policy(SET CMP0135 NEW)
 endif()
 
+if(WIN32)
+  if(NOT DEFINED BUILDENV_BASEPATH)
+    if(DEFINED ENV{BUILDENV_BASEPATH})
+      set(BUILDENV_BASEPATH "$ENV{BUILDENV_BASEPATH}")
+    else()
+      set(BUILDENV_BASEPATH "${CMAKE_SOURCE_DIR}/buildenv")
+    endif()
+  endif()
+
+  if(NOT DEFINED BUILDENV_URL)
+    if(DEFINED ENV{BUILDENV_URL})
+      set(BUILDENV_URL "$ENV{BUILDENV_URL}")
+    else()
+      message(STATUS "BUILDENV_URL not specified")
+    endif()
+  endif()
+
+  if(DEFINED BUILDENV_URL)
+    # Extract the filename from the URL
+    get_filename_component(FILENAME_WITH_EXTENSION "${BUILDENV_URL}" NAME)
+
+    # Remove the extension from the filename
+    # Note, that our buildenv file names contain two dots, e.g. mixxx-deps-2.5-x64-windows-875f13b.zip
+    string(
+      REGEX REPLACE
+      "\\.[^.]*$"
+      ""
+      BUILDENV_NAME
+      "${FILENAME_WITH_EXTENSION}"
+    )
+  else()
+    if(NOT DEFINED BUILDENV_NAME)
+      if(DEFINED ENV{BUILDENV_NAME})
+        set(BUILDENV_NAME "$ENV{BUILDENV_NAME}")
+      else()
+        message(
+          FATAL_ERROR
+          "Neither BUILDENV_URL nor BUILDENV_NAME specified!\nDid you run `${CMAKE_SOURCE_DIR}/tools/windows_buildenv.bat`?"
+        )
+      endif()
+    endif()
+  endif()
+
+  message(STATUS "BUILDENV_NAME is: ${BUILDENV_NAME}")
+
+  if(
+    EXISTS "${BUILDENV_BASEPATH}/${BUILDENV_NAME}"
+    AND IS_DIRECTORY "${BUILDENV_BASEPATH}/${BUILDENV_NAME}"
+  )
+    message(
+      STATUS
+      "BUILDENV directory ${BUILDENV_BASEPATH}/${BUILDENV_NAME} exists!"
+    )
+  else()
+    if(NOT DEFINED BUILDENV_URL)
+      message(
+        FATAL_ERROR
+        "BUILDENV_URL not specified!\nDid you run `${CMAKE_SOURCE_DIR}/tools/windows_buildenv.bat`?"
+      )
+    endif()
+    if(NOT DEFINED BUILDENV_SHA256)
+      if(DEFINED ENV{BUILDENV_SHA256})
+        set(BUILDENV_SHA256 "$ENV{BUILDENV_SHA256}")
+      else()
+        message(STATUS "BUILDENV_SHA256 not specified")
+      endif()
+    endif()
+
+    if(NOT EXISTS "${BUILDENV_BASEPATH}/${BUILDENV_NAME}.zip")
+      message(
+        STATUS
+        "Downloading file ${BUILDENV_URL} to ${BUILDENV_BASEPATH} ..."
+      )
+      file(
+        DOWNLOAD ${BUILDENV_URL} "${BUILDENV_BASEPATH}/${BUILDENV_NAME}.zip"
+        SHOW_PROGRESS
+        TLS_VERIFY ON
+      )
+    endif()
+
+    if(NOT ${BUILDENV_SHA256} STREQUAL "")
+      message(
+        STATUS
+        "Verify SHA256 of downloaded file ${BUILDENV_BASEPATH}/${BUILDENV_NAME}.zip ..."
+      )
+      file(SHA256 "${BUILDENV_BASEPATH}/${BUILDENV_NAME}.zip" actual_sha256)
+
+      if(NOT actual_sha256 STREQUAL ${BUILDENV_SHA256})
+        message(
+          FATAL_ERROR
+          "SHA256 checksum mismatch:\nexpected: ${BUILDENV_SHA256}\n     got: ${actual_sha256}"
+        )
+      else()
+        message(STATUS "SHA256 ${BUILDENV_SHA256} is correct!")
```

Comment:
Did you consider to join the initial message with this one: 
      message(
        STATUS
        "SHA256 checksum verified for downloaded file ${BUILDENV_BASEPATH}/${BUILDENV_NAME}.zip"
      )

and 
      message(
          FATAL_ERROR
          "SHA256 checksum mismatch for downloaded file ${BUILDENV_BASEPATH}/${BUILDENV_NAME}.zip:\nexpected: ${BUILDENV_SHA256}\n     got: ${actual_sha256}\nDownload may be corrupted. Please try again."
        )

Reply:
I would like to leave it as it is, as the one message tells what check is started, and another with result after. I think this is easier to understand than all in one line, and the user is always aware, what happens - this is a check that need 2-3 seconds.

Reply:
Ah OK that makes sense.